app/script/view_model/ConversationInputEmojiViewModel.js

```diff
@@ -0,0 +1,175 @@
+ * Copyright (C) 2017 Wire Swiss GmbH
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+'use strict';
+
+window.z = window.z || {};
+z.ViewModel = z.ViewModel || {};
+
+z.ViewModel.ConversationInputEmojiViewModel = class ConversationInputEmojiViewModel {
+  constructor() {
+    this.emoji_list = $('<div class="conversation-input-emoji-list" />');
+    this.emoji_dict = undefined;
+    this.emoji_start_pos = -1;
+
+    $.get('/image/emoji.tsv', (text) => {
+      this.emoji_dict = text.split('\n');
+    });
+  }
+
+  static get EMOJI_LIST_LENGTH() {
+    return 5;
+  }
+
+  static get MIN_QUERY_LENGTH() {
+    return 2;
+  }
+
+  on_input_key_down(data, event) {
+    if (this.emoji_list.is(':hidden')) {
+      return false;
+    }
+
+    switch (event.keyCode) {
+      case z.util.KEYCODE.ESC:
+        this.emoji_list.remove();
+        this.emoji_start_pos = -1;
+        break;
+      case z.util.KEYCODE.ARROW_UP:
+      case z.util.KEYCODE.ARROW_DOWN:
+        this.rotate_emoji_list(event.keyCode === z.util.KEYCODE.ARROW_UP);
+        this.suppress_key_up = true;
+        break;
+      case z.util.KEYCODE.ENTER: {
+        let input = event.target;
+        let emoji = this.emoji_list.find('>div.selected>span').html();
+        let val = input.value;
+        input.value = val.substr(0, this.emoji_start_pos - 1) + emoji + val.substr(input.selectionStart);
+        input.setSelectionRange(this.emoji_start_pos, this.emoji_start_pos);
+        this.emoji_list.remove();
+        this.emoji_start_pos = -1;
+        $(input).change();
+        break;
+      }
+      default:
+        return false;
+    }
+
+    event.preventDefault();
+    return true;
+  }
+
+  on_input_key_up(data, event) {
+    if (!this.suppress_key_up) {
+      let input = event.target;
+      let text = input.value || '';
+
+      if (text[input.selectionStart - 1] === ':') {
+        this.emoji_start_pos = input.selectionStart;
+        this.update_emoji_list(input);
+      } else if (this.emoji_start_pos !== -1) {
+        if (input.selectionStart < this.emoji_start_pos || text[this.emoji_start_pos - 1] !== ':') {
+          this.emoji_list.remove();
+          this.emoji_start_pos = -1;
+        } else {
+          this.update_emoji_list(input);
+        }
+      }
+    }
+
+    this.suppress_key_up = false;
+    return true;
+  }
+
+  update_emoji_list(input) {
```

Comment:
we used `snake_case` because of CoffeeScript (and other reasons long long time ago) and for sure we would like to change that since we are moving to ES6.. we'll have to confirm from the rest of the crew if they all (or majority) agree to use `camelCase` in the new modules..